surf/surf_test.go

```diff
@@ -0,0 +1,223 @@
+package surf
+
+import (
+	"bytes"
+	"fmt"
+	"math/rand"
+	"sort"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestBuildPrefixKeys(t *testing.T) {
+	buildAndCheckSuRF(t, [][]byte{
+		{1},
+		{1, 1},
+		{1, 1, 1},
+		{1, 1, 1, 1},
+		{2},
+		{2, 2},
+		{2, 2, 2},
+	})
+}
+
+func TestBuildCompressPath(t *testing.T) {
+	buildAndCheckSuRF(t, [][]byte{
+		{1, 1, 1},
+		{1, 1, 1, 2, 2},
+		{1, 1, 1, 2, 2, 2},
+		{1, 1, 1, 2, 2, 3},
+		{2, 1, 3},
+		{2, 2, 3},
+		{2, 3, 1, 1, 1, 1, 1, 1, 1},
+		{2, 3, 1, 1, 1, 2, 2, 2, 2},
+	})
+}
+
+func TestRandomKeysSparse(t *testing.T) {
+	buildAndCheckSuRF(t, randomKeys(2000000, 60, 0))
+}
+
+func TestRandomKeysPrefixGrowth(t *testing.T) {
+	buildAndCheckSuRF(t, randomKeys(100, 10, 200))
+}
+
+func TestBuildSuffixKeys(t *testing.T) {
+	buildAndCheckSuRF(t, [][]byte{
+		bytes.Repeat([]byte{1}, 30),
+		bytes.Repeat([]byte{2}, 30),
+		bytes.Repeat([]byte{3}, 30),
+		bytes.Repeat([]byte{4}, 30),
+	})
+}
+
+func TestMarshal(t *testing.T) {
+	keys := randomKeys(30, 20, 300)
+	vals := make([][]byte, len(keys))
+	for i := range keys {
+		vals[i] = make([]byte, 4)
+		endian.PutUint32(vals[i], uint32(i))
+	}
+	b := NewBuilder(4, 13, 13)
+	s1 := b.Build(keys, vals, 60)
+	var s2 SuRF
+	buf := s1.Marshal()
+	s2.Unmarshal(buf)
+	s1.checkEquals(t, &s2)
+	checkSuRF(t, &s2, keys, vals)
+}
+
+// max key length is `initLen * (round + 1)`
+// max result size is (initSize + initSize * (round + 1)) * (round + 1) / 2
+// you can use small round (0 is allowed) to generate a sparse key set,
+// or use a large round to generate a key set which has many common prefixes.
+func randomKeys(initSize, initLen, round int) [][]byte {
+	keys := make([][]byte, initSize)
+	rand := rand.New(rand.NewSource(0xdeadbeaf))
```

Comment:
Why use a static source?

Reply:
Easy for reproduce bug. Or we can dump keys when test failed?

Reply:
We can use a random source and print out the source before run the test.

Reply:
👍